systemd/systemd.go

```diff
@@ -931,57 +937,122 @@ func (s *systemd) IsActive(serviceName string) (bool, error) {
-	if _, err := s.systemctl(append([]string{"stop"}, serviceNames...)...); err != nil {
-		return err
-	}
 
-	// and now wait for it to actually stop
-	giveup := time.NewTimer(timeout)
-	notify := time.NewTicker(stopNotifyDelay)
-	defer notify.Stop()
-	check := time.NewTicker(stopCheckDelay)
-	defer check.Stop()
-
-	firstCheck := true
-loop:
-	for {
-		select {
-		case <-giveup.C:
-			break loop
-		case <-check.C:
-			allStopped := true
+	// Start the real time progress tracking inside a go routine because the
+	// 'systemctl stop' request is blocking (which runs in the parent function).
+	// Note that although the status polling thread is separate, and could start
+	// before the 'systemctl stop' command, the poll ticker will not fire
+	// immediately, allowing the stop command to start first. The ordering is
+	// not assumed, but it helps make existing unit-test code work, because the
+	// systemctl mocking in some modules are implemented very simplistically, and
+	// assumes that the 'stop' argument comes before the 'show'.
+	errorRet := make(chan error)
+	quit := make(chan interface{})
+
+	go func() {
+		// The polling routine is the 'errorRet' channel sender, so we make
+		// sure we exit closing the channel explicitly, even though we always
+		// return the error result first. Closing the channel does not free the
+		// object, the reader can still access the object. The object is only
+		// freed when no further references exit.
+		defer close(errorRet)
+
+		notify := time.NewTimer(stopNotifyDelay)
+		check := time.NewTicker(stopCheckDelay)
+		defer check.Stop()
+
+		// We start with a notice grace period to give services time to stop.
+		// Once this period expires, we provide user notifications every time
+		// the list of services we are waiting for changes (or on the first
+		// change of 'notifyEnable', which is detected by 'notifyShowFirst')
+		notifyEnable := false
+		notifyShowFirst := false
+
+		// Once the async systemd stop completes, we do one last status poll
+		// to make sure we have the most up to date state.
+		stopComplete := false
+
+		for {
+			select {
+			case <-quit:
+				// We are now complete, but poll final state of units
+				stopComplete = true
+
+			case <-check.C:
+				// Enable state poll of units
+
+			case <-notify.C:
+				// Enable user notifications and trigger first message
+				notifyEnable = true
+				notifyShowFirst = true
+			}
+
+			// Check if any of the remaining running units have stopped?
 			stillRunningServices := []string{}
 			for _, service := range serviceNames {
 				bs, err := s.systemctl("show", "--property=ActiveState", service)
 				if err != nil {
-					return err
+					errorRet <- err
+					return
 				}
 				if !isStopDone(bs) {
 					stillRunningServices = append(stillRunningServices, service)
-					allStopped = false
 				}
 			}
-			if allStopped {
-				return nil
+
+			// Any remaining running units stopped?
+			somethingChanged := len(serviceNames) != len(stillRunningServices)
```

Comment:
Thanks for this change, I was pondering about nothingChange before, this is more readable.

Reply:
Thanks @mardy for pointing out inverse logic is less fun ;-)